omnigibson/transition_rules.py

```diff
@@ -296,358 +710,453 @@ def transition(self, individual_objects, group_objects):
+                    continue
+
+                # Verify all required object quantities are contained in the container
+                if not cls._validate_recipe_objects_are_contained(recipe=recipe, in_volume=in_volume):
+                    continue
+
+                # Verify no non-relevant system is contained
+                if not cls._validate_nonrecipe_systems_not_contained(recipe=recipe, container=container):
+                    continue
+
+                # Verify no non-relevant object is contained
+                if not cls._validate_nonrecipe_objects_not_contained(recipe=recipe, in_volume=in_volume):
+                    continue
+
+                # Otherwise, all conditions met, we found a valid recipe and so we execute and terminate early
+                og.log.info(f"Executing recipe: {name} in container {container.name}!")
+
+                # Compute total volume of all contained items
+                volume = 0
+
+                # Remove all recipe system particles contained in the container
+                for system_name in recipe["input_systems"]:
+                    system = get_system(system_name=system_name)
+                    volume += contained_particles_state.get_value()[0] * np.pi * (system.particle_radius ** 3) * 4 / 3
+                    container.states[Filled].set_value(system, False)
+
+                # Remove all recipe objects
+                objs_to_remove = np.concatenate([
+                    cls._OBJECTS[np.where(in_volume[cls._CATEGORY_IDXS[obj_category]])[0]]
+                    for obj_category in recipe["input_objects"].keys()
+                ]).tolist()
+                volume += sum(obj.volume for obj in objs_to_remove)
+                t_results.remove += objs_to_remove
+
+                # Spawn in new fluid
+                out_system = get_system(recipe["output_system"])
+                out_system.generate_particles_from_link(
+                    obj=container,
+                    link=contained_particles_state.link,
+                    mesh_name_prefixes="container",
+                    max_samples=volume // (np.pi * (out_system.particle_radius ** 3) * 4 / 3),
+                )
+
+                # Terminate early
+                return t_results
+
+            # Otherwise, if we didn't find a valid recipe, we execute a garbage transition instead
+
+            # Remove all systems inside the container
+            for system in PhysicalParticleSystem.get_active_systems():
+                if container.states[Contains].get_value(system):
+                    container.states[Filled].set_value(system, False)
+
+            # Remove all objects inside the container
+            objs_to_remove = cls._OBJECTS[np.where(in_volume)[0]].tolist()
+            t_results.remove += objs_to_remove
+
+            # Spawn in garbage fluid
+            container.states[Filled].set_value(get_system(system_name=m.DEFAULT_GARBAGE_SYSTEM), True)
```

Comment:
nit: is it possible to include this as an actual recipe and we intentionally place this recipe in the last element of `cls._ACTIVE_RECIPES`? Then all the code above can be reused.

Reply:
generate correct amount of garbage. maybe helper function?